bfvmm/src/hve/arch/intel_x64/vmexit/cpuid.cpp

```diff
@@ -227,21 +224,20 @@ execute_emulators(vcpu *vcpu, const std::list<handler_delegate_t> &emulators)
 bool
 cpuid_handler::handle(vcpu *vcpu)
 {
-    const auto &emulators =
-        m_emulators.find(vcpu->rax());
+    vcpu->set_gr1(vcpu->rax());
+    vcpu->set_gr2(vcpu->rcx());
```

Comment:
this is good. The only thing that this is missing is in the emulation code, we need to set rax, rbx, rcx, and rdx to 0 now that you have fixed this bug.

Reply:
Which scenario should rax/rbx/rcx/rdx always be zeroed out? Does Intel specify that CPUID returns 0 for all leafs that don't exist?

Reply:
Yes, 0 is defined as the return value for CPUID. The emulation code should zero out the registers just before this line:
https://github.com/Bareflank/hypervisor/blob/e3b641225ead1fee2b2c3bdeb2c32d84e4265f44/bfvmm/src/hve/arch/intel_x64/vmexit/cpuid.cpp#L215

Reply:
We might need to think a little more about this one... I thought the purpose of the "emulator" path was to allow extension writers an un-tampered-with path to take over handing of an event. Zeroing out the output registers is slightly tampering with the handling of the event.

Reply:
huh... good point. this comment got me thinking. In the MSR case, your 100% right. In the event that an MSR is not supported, the proper response is to leave the registers alone and inject a GPF. I'll mark this as resolved. If you emulate, you should zero out the registers manually. It's a bit of extra work, but there really is no clean way to know what to do in each case, so let the delegate figure it out.